### PR TITLE
Merged discussions to RFC1004

### PR DESCRIPTION
**Pull request by tailhook**
* In particular "Failover and Replication Pre-RFC"
* And "Stateful Connection Configuration Pre-RFC"

### CODE REVIEW

File: text/1004-transactions-api.rst [hunk 1]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+field is present) as it indicates that query has been analyzed.
 
+The ``SERVER_OPT_HAS_FEATURES`` is needed for the following tasks:
 
-Retry Single Queries on Connection Error
-----------------------------------------
+1. Retry standalone (non-transactional) read-only queries
+2. Warn when features are used in inapropriate context (warn when
+   session modification queries are sent on non-raw connection, which
+   means they can be lost at any point due to reconnect)
```
Comment (1st1) [inline text/1004-transactions-api.rst:704]: Hm, how will the server know if the connection is raw or non-raw? IMO it shouldn't care, it's entirely up to the client.

File: text/1004-transactions-api.rst [hunk 2]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
-single mutable request on a "concurrent update" error must be a
-deliberate decision.
+``SERVER_OPT_HAS_FEATURES`` is an empty string for read-only queries (the
+field is present) as it indicates that query has been analyzed.
 
+The ``SERVER_OPT_HAS_FEATURES`` is needed for the following tasks:
 
-Retry Single Queries on Connection Error
-----------------------------------------
+1. Retry standalone (non-transactional) read-only queries
```
Comment (1st1) [inline text/1004-transactions-api.rst:701]: You mean that the server would retry some of the client's queries on its own? I wouldn't like that.

File: text/1004-transactions-api.rst [hunk 3]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+===========
 
-To make changing EdgeDB address or restarting EdgeDB work nicely, we
-need to retry simple queries on connection errors too:
+
+More Configuration
+------------------
+
+Setting tracing/debugging metadata for queries may be implemented
+in the same way:
```
Comment (1st1) [inline text/1004-transactions-api.rst:737]:
```suggestion
in the same way in the future:
```

File: text/1004-transactions-api.rst [hunk 4]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
-4. Or vice versa add ``pool.read_only_query()`` method
-5. Add ``pool.read_only().query()``
-6. Add ``pool.with_options({readonly: true}).query()``
+   def read_only(self, primary=False, max_replica_lag=10): ...
 
-Note: methods (2-6) are also helpful for working with primary/replica
-installations. But probably only last two would allow full power, as they allow
-``pool.read_only(primary=true)`` (i.e. in case you need read-only transaction
-that can't go to a replica).
+Perhaps this should be encapsulated into replica options:
```
Comment (1st1) [inline text/1004-transactions-api.rst:770]: Let's not use "perhaps" here -- we need to decide now or leave this for future RFCs

File: text/1004-transactions-api.rst [hunk 5]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+.. code-block:: python
+
+   def read_only(self, primary=False, replicas: ReplicaOptions): ...
 
 
 Learning Curve
 ==============
 
-This complicates learning curve, but:
+The ``retry`` method complicates learning curve curve, but:
```
Comment (1st1) [inline text/1004-transactions-api.rst:780]:
```suggestion
The ``retry`` method complicates the learning curve, but:
```

File: text/1004-transactions-api.rst [hunk 6]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
 
 1. This is already a problem in Postgres, there a lot of people who
-   ignore the issue for pet projects and a lot of startups having "normal"
-   rate or 500 errors at any point of time, while it's preventable.
+   ignore the issue for pet projects and a lot of startups having
+   "normal" rate or 500 errors at any point of time, while it's
+   preventable.
```
Comment (1st1) [inline text/1004-transactions-api.rst:785]: "1. Letting the application to error out instead of automatically retrying certain transactions is a wildly known, yet an entirely preventable problem."

File: text/1004-transactions-api.rst [hunk 7]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
 2. Repeating transactions would make less incentive to keep transactions
-   open for a long time (e.g. while accessing slow network resources like
-   external API), which is a problem of itself.
+   open for a long time (e.g. while accessing slow network resources
+   like external API), which is a problem of itself.
```
Comment (1st1) [inline text/1004-transactions-api.rst:788]: "2. The design of the proposed `retry()` method emphasizes that the code block might be executed more than one time, suggesting to the user to factor out slow blocking code, like making API calls over network. This ensures that DB transactions would not be open longer than it is necessary."

File: text/1004-transactions-api.rst [hunk 8]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
 3. Even if we never have failed concurrent updates we would want
    seamless reconnect on connection failures (i.e. server restart,
    primary/replica change, etc.)
```
Comment (1st1) [inline text/1004-transactions-api.rst:791]: "3. The API adds resilience against not only errors related to failing to serialize concurrent updates, but also against possible network errors caused by the DB server restart, primary/replica changes, etc."

File: text/1004-transactions-api.rst [hunk 9]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
 4. To make learning curves shorter I think we should intentionally
-   inject failures. This is needed so that users quickly find out that side
-   effects of their transactions are in effect several times.
+   inject failures. This is needed so that users quickly find out that
+   side effects of their transactions are in effect several times.
```
Comment (1st1) [inline text/1004-transactions-api.rst:794]: "4. In the future we will consider adding a connection setting to inject failures into transactions to help users discover incorrect transactional code."

@tailhook @elprans I think we shouldn't do this before 1.0 as I'm not sure how it might impact benchmarks, production code, and the overall DX of early adopters. I think that our API is already going to be structured in the right way, and adding failure injection is entirely optional at this point.

File: text/1004-transactions-api.rst [hunk 10]
```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
    network failures would be an issue
 
 
-Keep `transaction` as but Add a Helper for Retrying
----------------------------------------------------
+Keep `transaction` as is but Add a Helper for Retrying
```
Comment (1st1) [inline text/1004-transactions-api.rst:948]:
```suggestion
Separate Retry API
```

File: text/1004-transactions-api.rst [hunk 11]
```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
...
+------------------------------------------------------
 
-The problem with this approach is that it hard to teach using `retry`
+The problem with this approach is that it hard to teach using ``retry``
```
Comment (1st1) [inline text/1004-transactions-api.rst:951]: "The problem of adding a separate *transaction retry API* while keeping the current `Connection.transaction()` family of methods is that it will be hard to educate the users that they should almost always use the former. Most of the time transaction serialization and network errors are uncommon during the development and only surface when in production."

File: text/1004-transactions-api.rst [hunk 12]
```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
...
 
 
+Retry All Single Queries
+------------------------
+
+This specification describes that read-only non-transactional queries
+should be retried automatically.
+
+While it's tempting to retry all ``pool.query`` and ``pool.execute``
+calls (even modifications), it **gives the false sense of security**: no
```
Comment (1st1) [inline text/1004-transactions-api.rst:963]:
```suggestion
calls (even those that execute queries with side effects), it **gives the false sense of security**: no
```

File: text/1004-transactions-api.rst [hunk 13]
```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
...
+"concurrent update" issues are seen.  But it's better to see such error
+and turn the whole block of code into a transaction rather than just a
+mutation. I.e. retrying a single mutable request on a "concurrent
+update" error must be a deliberate decision.
```
Comment (1st1) [inline text/1004-transactions-api.rst:967]: This whole paragraph is very hard to parse. I suggest to reword it to:

"Automatically retrying `query()`-like calls that execute queries with side effects is not always safe. Therefore our recommendation to users would be to use the new `retry()` API when they know that the query would be safe to repeat."

File: text/1004-transactions-api.rst [hunk 14]
```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
...
+
+
+Is ``with_modifications`` Needed?
+---------------------------------
+
+It's intuitive that databases are mutable by default. But there is a
+large class of applications that are mostly read-only and must have
+limited and easy to find places having mutations. For those apps it's
+better to have read-only connection by default and use a pattern like
+this for writes:
```
Comment (1st1) [inline text/1004-transactions-api.rst:977]: The RFC starts with stating and showing examples of `edgedb.connect()` returning a mutable connection. So I assume that this section advocates for adding a way to convert a read-only connection (obtained via `edgedb.connect().read_only()`) to a mutating one?  I'm confused.

File: text/1004-transactions-api.rst [hunk 15]
```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
...
+configuration anyway. So it deviates from the usual behavior of context
+manager anyway.
+
+
+Disabling Features
+------------------
+
+We may introduce a pool and/or connection configuration to disable
+``ddl`` and ``config`` features on the requests. And/or disable
+``session`` and ``transaction`` features on connection. This should be
```
Comment (1st1) [inline text/1004-transactions-api.rst:1058]: I like the disabling idea. I'm not sure if it should be the default configuration though, we can just teach that it's a good pattern to narrow down the connection functionality to what your app needs.

**Review by 1st1 — COMMENTED** [review]
Wow. Overall this looks really close. Thanks for championing this and for spending time to convince us that these ideas are the best path forward.

I've left some comments with questions and edit suggestions.  I can do a heavy editing pass once this is merged.

File: text/1004-transactions-api.rst (hunk 9 above)
Comment (elprans) [reply text/1004-transactions-api.rst:794]: I also think that proper API design, documentation and tutorials would be sufficient to guide users toward the correct set of idioms.  Intentional failure injection might be an optional feature.

File: text/1004-transactions-api.rst (hunk 15 above)
Comment (elprans) [reply text/1004-transactions-api.rst:1058]: Agree.  I think we should optimize for simplicity at a glance, in which case lengthy connection "setups" in simple tutorials would scare users away.

**Review by elprans — COMMENTED** [review]
Great work, @tailhook!

File: text/1004-transactions-api.rst [hunk 16]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
-tempting to add a method on connection object itself:
+This RFC introduces a connection parameter:
 
 .. code-block:: python
 
-    with pool.acquire() as conn:
-        for tx in conn.retry():
-            with tx:
-                tx.query(..)
+    edgedb.connect('inst1', wait_until_available_sec=30)
```
[the remark this answers is not in the document]
Comment (1st1) [reply text/1004-transactions-api.rst:623]: Hard and solid -1 from me. One library shouldn't fight the language for unclear benefits.

File: text/1004-transactions-api.rst [hunk 17]
```diff
@@ -68,6 +89,170 @@ to handle:
...
+* Setting session configuration
+* Transaction options
+
+Here are method signatures:
+
+.. code-block:: python
+
+    class Connection:
+        ...
+        def read_only(self,
```
[the remark this answers is not in the document]
Comment (1st1) [reply text/1004-transactions-api.rst:178]: I think `as_*` is more common for this kind of API.

File: text/1004-transactions-api.rst [hunk 18]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
 
+In case of ``SERVER_OPT_HAS_FEATURES`` it describes what is actually
+contained in the query. And in case of ``QUERY_OPT_ALLOW_FEATURES``
+client can specify what of these things are allowed in this query.
 
-Retry Single Queries
---------------------
+Read-only queries are always allowed. When ``QUERY_OPT_ALLOW_FEATURES``
+is omitted any query is allowed (default). With the empty string only
+read-only queries are allowed.
```
[the remark this answers is not in the document]
Comment (1st1) [reply text/1004-transactions-api.rst:694]: I don't understand why we need this complexity.  IMO instead of "modification" header flag we should have "readonly". If it's set the query fails to prepare unless it's readonly. This is way easier to understand and document.

File: text/1004-transactions-api.rst (hunk 14 above)
Comment (1st1) [reply text/1004-transactions-api.rst:977]: Yeah. Paul, we've argued over this a lot already, and I see no end to this. So I propose we do this instead: we don't implement `with_modifications()` right now. We can always add this later, it's really not a critical feature and we can add it at any time when we get some user feedback.

I'd really like this RFC to only implement essential and **non-controversial** APIs, where non-controversial means unanimous support of you, Elvis, and I.

File: text/1004-transactions-api.rst [hunk 19]
```diff
@@ -68,6 +89,170 @@ to handle:
...
+1. They send read-only flag to the compiler, so compiler rejects queries
+   that modify data. And *may* use replica connection.
+2. Works as type-checker hint that function that received a connection
+   does no modifications to the data.
+
+``RawConnection`` class is exposed for users want to do their own
+session config management stuff, connection pooling or control of
+reconnection.
+
+Note: this section contains only sync Python example, async Python and
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:155]: Oh, well, forgot that entirely. My idea was that `AsyncIO` prefixes and `async` in signatures don't add any value, but...

File: text/1004-transactions-api.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:178]: Then we have to change all other methods. But `as_session_config` is not the same as `with_session_config`. So let's make `with_read_only` then?

File: text/1004-transactions-api.rst [hunk 20]
```diff
@@ -68,6 +89,170 @@ to handle:
...
+
+.. code-block:: python
+
+    conn = edgedb.connect()
+    read_only_conn = conn.read_only()
+    conn.execute("INSERT User { ... }", ...)
+    print(read_only_conn.query("SELECT User"))
+    read_only_conn.execute("INSERT User { .. }", ...)  # throws an error
+    conn.read_only(allow_upgrade=True) \
+        .with_modifications().execute("INSERT User { ... }", ...)  # good
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:211]: Well, I still think this is very useful pattern that I would use in **every** my project. But if we have to cut it to do at least one step forward, okay.

File: text/1004-transactions-api.rst [hunk 21]
```diff
@@ -224,7 +416,7 @@ All network error within connection should be converted into
 ``EarlyNetworkError`` or ``NetworkError``. Former is used in context
 where we catch network error before sending a request.
 
-Additionally ``TransactionActiveError`` is introduced to signal that
+``TransactionActiveError`` is introduced to signal that
 queries can't be executed on the connection object:
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:420]: Added explanation

File: text/1004-transactions-api.rst [hunk 22]
```diff
@@ -224,7 +416,7 @@ All network error within connection should be converted into
 ``EarlyNetworkError`` or ``NetworkError``. Former is used in context
 where we catch network error before sending a request.
 
-Additionally ``TransactionActiveError`` is introduced to signal that
+``TransactionActiveError`` is introduced to signal that
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:419]: `TransactionIsActiveError`? 

`ActiveTransactionError` -- reads as error of active transaction (i.e. some base for a whole class of errors).

Also added clarification below.

File: text/1004-transactions-api.rst [hunk 23]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
-Future Work
-===========
+Server Availability Timeout
+---------------------------
 
-Transaction on Specific Connection
-----------------------------------
+Previously, when TCP connect gets "connection refused" error or when
+timeout happens on handshake both connection and pool API would crash
+the application. This is weird when initially starting an application
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst]: Changed to "can be inconvenient". This isn't particularly confusing.

File: text/1004-transactions-api.rst [hunk 24]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+3. Connection reset, connection aborted, connection refused (server is
+   restarting or not ready yet)
+4. Timeout happened during connect or authentication
+
+All other errors are propagated immediately.
+
+This timeout is different from ``connect_timeout``. Connect timeout
+determines how long individual connect attempt may take. And
+``wait_until_available`` determines how many such attempts could be
+made (i.e. how many ones fits the time frame).
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:667]: Previously we discussed that at least with @1st1 and he was for this change.

Here are my points:
1. For nice `docker-compose up` we need `wait_until_available` = 30 sec
2. Current python timeout is 60 sec. So by introducing `wait_until_available` 30 sec we don't make application hanging longer than what was there. In particular, I propose to lower `connect_timeout` to 10 seconds (for a single connection attempt). With connection pooling (ports RFC) this should work well enough.
3. If default is `0` nobody would ever discover `wait_until_available` and will put their own hacks for running locally, like they do with postgres. Unless we put this setting in every tutorial (which we definitely don't want for conciseness).
4. We can't crash app on every `query()` that got reconnected and caught `ConnectionRefused` (and get app restarted). So it's also a bunch of 500 errors on EdgeDB restart or failover.

*(the case when app and database starts simultaneously with the database on kubernetes is so much rare that we don't care about that use case)*

File: text/1004-transactions-api.rst [hunk 25]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+EdgeDB Changes
+--------------
+
+To support features above we add two headers to EdgeDB queries:
 
-While this covers "concurent update" and "deadlock" errors, with the
-current implementation it can't reconnect when when connection was
-interrupted while transaction is active. There are few ways it can work:
+1. For PrepareComplete_, CommandComplete_ server-side messages:
+   ``SERVER_OPT_HAS_FEATURES``
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:676]: It's kinda just symmetry with ``QUERY_OPT``. ``SERVER_HAS_FEATURES`` reads bad. But better name is appreciated.

File: text/1004-transactions-api.rst (hunk 2 above)
Comment (tailhook) [reply text/1004-transactions-api.rst:702]: Not server, client.

File: text/1004-transactions-api.rst [hunk 26]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+1. `modification` -- query is not read-only
+2. `session` -- query contains session config change
+3. `config` -- server or database config change
+4. `ddl` -- query contains DDL
+5. `transaction` -- query contains start/commit/rollback of transaction
+    or savepoint manipulation
 
+In case of ``SERVER_OPT_HAS_FEATURES`` it describes what is actually
+contained in the query. And in case of ``QUERY_OPT_ALLOW_FEATURES``
+client can specify what of these things are allowed in this query.
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:691]: Not sure I understand. It's in the next paragraph.

File: text/1004-transactions-api.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:695]: This is because:
```
conn.execute("""
START TRANSACTION;
UPDATE User SET {...}
COMMIT
""")
```
Would return `transaction,modifications` flag.

With @1st1 proposal `.execute("START TRANSACTION")` would return `transaction,readonly` features and `.execute("START TRANSACTION; UPDATE..."` would return `transaction`.

I.e. purely additive features are better at least for `HAS_FEATURES`, but is not bad for `ALLOW` too.

I we go with bitmask read-only queries == '0' is also good enough.

File: text/1004-transactions-api.rst (hunk 1 above)
Comment (tailhook) [reply text/1004-transactions-api.rst:705]: Here is the strawman (but all of these are implementation details):
```
class RawConnection:
    def _query_with_options(self, options, query, *args, **kw): ...
    def query(self, query, *args, **kw):
        self._query_with_options({'allow': '*'}, query, *args, **kw)

class Connection:
   def query(self, query, *args, **kw):
       self._raw._query_with_options({'warn': ['transaction','session_config']}, query, *args, **kw)

class ReadonlyConnection:
   def query(self, query, *args, **kw):
       self._raw._query_with_options({'allow': []}, query, *args, **kw)

class PoolConnectionProxy:
   def query(self, query, *args, **kw):
       self._raw._query_with_options({'deny': ['transaction','session_config']}, query, *args, **kw)

class Transaction:
    def __enter__(self):
          self._conn._raw._query_with_options({'allow': ['transaction']}, "START TRANSACTION")
    def query(self, *args, **kw):
          self._conn._raw._query_with_options({'allow': ['modifications', 'ddl']}, query, *args, **kw)
````
(these allow/deny are to show the concept, we're probably going to have a constant with the bit mask already prepared for the specific use case for efficiency)

File: text/1004-transactions-api.rst [hunk 27]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+Future Work
+===========
 
-To make changing EdgeDB address or restarting EdgeDB work nicely, we
-need to retry simple queries on connection errors too:
+
+More Configuration
+------------------
+
+Setting tracing/debugging metadata for queries may be implemented
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:737]: That is a future work. Tags are essentially a specific way to organize metadata. Also OpenTracing is deprecated, and OpenTelemetry should be used instead.

Also `with_tags` has little meaning. `with_telemetry_tags` or `with_tracing_tags` looks better, but let's not bikeshed about this here.

File: text/1004-transactions-api.rst (hunk 4 above)
Comment (tailhook) [reply text/1004-transactions-api.rst:771]: This is a "Future Work" section. What's wrong in "perhaps" here? This just gives more context to the current decisions but don't have to be clearly defined in this RFC.

File: text/1004-transactions-api.rst (hunk 9 above)
Comment (tailhook) [reply text/1004-transactions-api.rst:796]: This is where we constantly disagree. Documentation and tutorials are not enough for the same reason for which we use unit tests to test code. As we now basically know the distinction between workstation installations (created by `edgedb server init`) and production (roughly -- everything else), it's possible to have server itself without failure injection by default, but `server init` configuring it in a way that enables this by default. Also that section describes a way of failure injection that doesn't impact benchmarks in any significant way.

But anyway, this RFC is already hanging for too long, so I drop the "failure injection" section in favor of later RFC.

File: text/1004-transactions-api.rst (hunk 13 above)
Comment (tailhook) [reply text/1004-transactions-api.rst]: This is not enough in my opinion. Added an example for clarification.

File: text/1004-transactions-api.rst (hunk 14 above)
Comment (tailhook) [reply text/1004-transactions-api.rst:978]: > So I assume that this section advocates for adding a way to convert a read-only connection (obtained via edgedb.connect().read_only()) to a mutating one?

Yes, in the top of the application, you either do (assuming `aiohttp`):
```python
app['edgedb'] = await create_async_pool('name')
```
Or you do:
```python
app['edgedb'] = (await create_async_pool('name')).read_only(allow_upgrade=True)
```
In the latter case, you prevent arbitrary functions execute side-effects unless run within `with_modification().retry()` transaction and transaction is passed to the function directly.

> In my mind, the main feature and premise of read_only() is that it allows better scalability by allowing connections to standby replicas.

Well:
1. We don't implement that scalability just yet anyway
2. With regards of this RFC the original premise of read-only was retrying queries automatically. But well, it's solved by `HAS_FEATURES` so is not strictly needed
3. I don't see anything in `with_modifications` as controversial (except the name). Controversial would be official recommendation of this, but not the support. And in previous discussion I think it was agreed as "yes we can support this pattern".

Anyway I'm dropping the function and move this section to the Future Work, so that we can finally move forward.

File: text/1004-transactions-api.rst (hunk 15 above)
Comment (tailhook) [reply text/1004-transactions-api.rst:1059]: Looks like you agree with what is written in the RFC. Resolving the conversation.

File: text/1004-transactions-api.rst [hunk 28]
```diff
@@ -39,7 +60,7 @@ errors.
 
 `More discussion on this topic <https://github.com/edgedb/edgedb/discussions/1708>`_
 
-Second part of the RFC is mainly motiveated by the fact that PostgreSQL
+This part of API change is also motivated by the fact that PostgreSQL
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst:63]: My point was that unprepared reader can read this and say that "edgedb is less stable and throws more errors than postgres and that all this transaction repeating thing is a workaround to handle this", which isn't true. Another backend could tackle this issue differently (i.e. locking everything). And we generally don't hide that we're using postgres as a backend.

With all that said should I replace Postgres to EdgeDB?

File: text/1004-transactions-api.rst [hunk 29]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
-b. It may reconnect and replace underlying socket in the connection
-   object and retry
-c. We may only allow transactions on connection pools (current
-   specification)
+Both contain the set of strings joined by comma:
 
-This can be postponed to later RFC when we know what are use cases of
-it.
+1. `modification` -- query is not read-only
+2. `session` -- query contains session config change
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst]: Changed and also use bitmap, so names matter much less.

File: text/1004-transactions-api.rst [hunk 30]
```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
...
+3. `config` -- server or database config change
+4. `ddl` -- query contains DDL
+5. `transaction` -- query contains start/commit/rollback of transaction
+    or savepoint manipulation
```
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst]: I don't think that performance here would be noticeable, but bitmap sounds good anyway. and using all ones to "allow everything" is good too.

File: text/1004-transactions-api.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (tailhook) [reply text/1004-transactions-api.rst]: Fixed

File: text/1004-transactions-api.rst [hunk 31]
```diff
@@ -667,24 +667,24 @@ To support features above we add two headers to EdgeDB queries:
 2. For Prepare_, OptimisticExecute_, Execute_, ExecuteScript_
    client-side messages: ``QUERY_OPT_ALLOW_FEATURES``
 
-Both contain the set of strings joined by comma:
+Both contain 32bit bitmap of the following:
```
Comment (elprans) [inline text/1004-transactions-api.rst]: Let's do 64 bits.  No reason to conserve space here.

File: text/1004-transactions-api.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (elprans) [reply text/1004-transactions-api.rst]: I agree with @tailhook.  It's easier to reason when each bit _extends_ capabilities consistently as opposed to either restricting or extending them.

File: text/1004-transactions-api.rst (hunk 28 above)
[the remark this answers is not in the document]
Comment (elprans) [reply text/1004-transactions-api.rst:63]: I think writing it as "EdgeDB (as is PostgreSQL)" strikes a good balance between your and Yury's concern.
